lctrie_ip.c

```diff
@@ -79,13 +79,9 @@ size_t subnet_dedup(lct_subnet_t *subnets, size_t size) {
       // assume that the prior defined subnet is the desired one,
       // dis-allowing redefinition of that subnet elsewhere,
       // ex. bogon file, BGP ASN list, user specified subnets
-      //
-      // slide the rest of the array over the second value.  if we're at the
-      // end of the array, just let it drop off.
-      if ((j + 1) < size)
-        memmove(&subnets[j], &subnets[j + 1], (size - (j + 1)) * sizeof(lct_subnet_t));
-      --size;
       ++ndup;
+    } else {
+      subnets[++i] = subnets[j];
```

Comment:
This is the case that the two consecutive elements are not duplicates and you added an assignment of the element at ++i (this is prefix operator and will return you the value after increment, so not even element at i) to the element at j.  Why?  This doesn't make any sense to me and will absolutely wreck the pre-sorted array ordering.

Reply:
Initially i = 0 and j = 1 -- i represents the current end of the deduplicated subnets and j represents the current subnet being inspected.

We compare subnet at i with the subnet at j, since if j was a duplicate then they would be equal (and so would all the elements between i + 1 and j - 1)

And if subnet at j is not a duplicate, you increment i before assigning j -- this is the new end of the deduplicated array

Reply:
Essentially, the first element alone is a deduplicated array, and we are appending unique elements to the end of it

Reply:
Thus we are never swapping subnets since i < j (hence i + 1 <= j) at all times

What this code achieves is to build the deduplicated array (still in sorted order) in-place by swapping over unique elements to the end each time